bootstrap/kubeadm/controllers/kubeadmconfig_controller.go

```diff
@@ -441,13 +440,9 @@ func (r *KubeadmConfigReconciler) joinWorker(ctx context.Context, scope *Scope)
 	}
 	conditions.MarkTrue(scope.Config, bootstrapv1.CertificatesAvailableCondition)
 
-	// ensure that joinConfiguration.Discovery is properly set for joining node on the current cluster
-	if err := r.reconcileDiscovery(ctx, scope.Cluster, scope.Config, certificates); err != nil {
-		if requeueErr, ok := errors.Cause(err).(capierrors.HasRequeueAfterError); ok {
-			scope.Info(err.Error())
-			return ctrl.Result{RequeueAfter: requeueErr.GetRequeueAfter()}, nil
-		}
-		return ctrl.Result{}, err
+	// Ensure that joinConfiguration.Discovery is properly set for joining node on the current cluster.
+	if res, err := r.reconcileDiscovery(ctx, scope.Cluster, scope.Config, certificates); err != nil || !res.IsZero() {
+		return res, err
 	}
```

Comment:
This feels a bit odd given the convention of not relying on other values when an error occurs, might be better to be explicit here:
```suggestion
	res, err := r.reconcileDiscovery(ctx, scope.Cluster, scope.Config, certificates)
	if err != nil  {
		return ctrl.Result{}, err
	}
	if !res.IsZero() {
		return res, nil
	}
```

Reply:
I like that!

Reply:
Fixed it up, although I kept the block with an else to keep the variables scoped, wdyt?

Reply:
Works for me, @ncdc is the one that despises `else` :joy: